lib/THC/THCReduceApplyUtils.cuh

```diff
@@ -72,6 +72,32 @@ __device__ T reduceBlock(T* smem,
   return r;
 }
 
+// Block-wide reduction where each thread locally reduces N
+// values before letting a single warp take over
+template <typename T, typename ReduceOp, int N, bool SyncAfterLocalReduction>
+__device__ T reduceBlockN(T *smem,
+                         T threadVals[N],
+                         int numVals,
+                         ReduceOp reduceOp,
+                         T init) {
+  int offset = threadIdx.x * N;
+  T local = offset < numVals ? threadVals[0] : init;
+
+#pragma unroll
+  for (int i = 1; i < N; ++i) {
+    ++offset;
+    T next = offset < numVals ? threadVals[i] : init;
+    local = reduceOp(local, next);
+  }
+
+  // Needed if threadVals points to the same shared memory as smem
+  if (SyncAfterLocalReduction) {
```

Comment:
I think the idea is that threadVals is always in registers, not smem, thus this is not needed

(just as `T threadVal` for `reduceBlock` is always in a register)